phosphor-auditlog/alog_parser.hpp

```diff
@@ -0,0 +1,258 @@
+#pragma once
+
+#include "alog_utils.hpp"
+
+#include <auparse.h>
+#include <libaudit.h>
+
+#include <nlohmann/json.hpp>
+#include <phosphor-logging/lg2.hpp>
+#include <sdbusplus/bus.hpp>
+#include <sdbusplus/server/object.hpp>
+#include <xyz/openbmc_project/Common/File/error.hpp>
+#include <xyz/openbmc_project/Common/error.hpp>
+#include <xyz/openbmc_project/Logging/AuditLog/server.hpp>
+
+#include <fstream>
+#include <list>
+#include <string>
+
+namespace phosphor::auditlog
+{
+
+/** @class ALParser
+ *  @brief Parsing audit log using auparse library services
+ *  @details Provides abstraction to auparse library services
+ */
+class ALParser
+{
+  public:
+    ALParser(const ALParser&) = delete;
+    ALParser& operator=(const ALParser&) = delete;
+    ALParser(ALParser&&) = delete;
+    ALParser& operator=(ALParser&&) = delete;
+
+    /** @brief Constructor to initialize parsing of audit log files
+     *  @details Prepares parsedFile for writing of audit events
+     *  @param[in] parsedFile Initialized file for holding parsed log events
+     */
+    ALParser(ALParseFile& parsedFile)
+    {
+        if (!openParsedFile(parsedFile.getPath()))
+        {
+            throw sdbusplus::xyz::openbmc_project::Common::File::Error::Write();
+        }
+    }
+
+    ~ALParser()
+    {
+        auparse_destroy(au);
+    }
+
+    /**
+     * @brief Process audit events from initialized au source
+     */
+    virtual void doParse();
+
+    /**
+     * @brief Process audit events from initialized au source
+     */
+    void processEvents();
+
+    /**
+     * @brief Format audit entries into raw JSON
+     * @param[in,out] parsedEntry Filled in with parsed audit entry.
+     * @return bool True if parsing succeeded, false otherwise.
+     */
+    virtual bool formatEntry(nlohmann::json& parsedEntry)
+    {
+        return formatRaw(parsedEntry);
+    };
+
+    /**
+     * @brief Formats next record into JSON format using message registry
+     * @param[in,out] parsedEntry Filled in with parsed audit entry.
+     * @return bool True if parsing succeeded, false otherwise.
+     */
+    bool formatMsgReg(nlohmann::json& parsedEntry);
+
+    /**
+     * @brief Formats next record into JSON general format
+     * @param[in,out] parsedEntry Filled in with parsed audit entry.
+     * @return bool True if parsing succeeded, false otherwise.
+     */
+    bool formatGen(nlohmann::json& parsedEntry);
+
+    /**
+     * @brief Formats next record into JSON raw format
+     * @param[in,out] parsedEntry Filled in with parsed audit entry.
+     * @return bool True if parsing succeeded, false otherwise.
+     */
+    bool formatRaw(nlohmann::json& parsedEntry);
+
+    auparse_state_t* au = nullptr;
+    std::ofstream parsedStream;
+
+  private:
+    /**
+     * @brief Moves parser to point to next event
+     *
+     * @return false when no more events exist, or on error
+     */
+    bool auditNextEvent();
```

Comment:
nit: This sounds like it will be auditing the next event.  How about something like 'parseNextEvent' or 'moveToNextEvent' or something like that?

Reply:
Good idea, I've renamed to `getNextEvent()`